File: CryptoAC/src/main/test/eu/fbk/st/cryptoac/util/SafeRegexTest.java
package eu.fbk.st.cryptoac.util;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;


/**
 * This class contains methods to test RegExs for validating users' input through.
 * As we took RegEx from OWASP, we assume they were thoroughly tested. Therefore,
 * here we just check some basic properties.
 */
class SafeRegexTest {

    // These are the regular expressions to test
    // - safeTextRegex = "^[-a-zA-Z0-9._/+=]+$";
    // - urlRegex = "^(https?)://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]$";
    // - uriRegex = "^[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]$";
    // - ipv4Regex = "^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])(\\.(?!$)|$)){4}$";
    // - urlOrIPRegex = urlRegex + "|" + ipv4Regex;
    // - base64Regex = "^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$";
    // - awsAccessKeyRegex = "^[A-Z0-9]{20}$";
    // - awsSecretKeyRegex = "^[A-Za-z0-9/+=]{40}$";

    String exampleSafeTextNoSpacesRegex = "text";
    String exampleUrlRegex = "https://google.com/";
    String exampleUriRegex = "/#G1kiGui2qDYcdjriD0kdsspLycD75of";
    String exampleIpv4Regex = "192.168.111.4";
    String exampleBase64Regex = "At539hww";
    String exampleAwsAccessKeyRegex = "A94KF94IR9FIEKDMCJ84";
    String exampleAwsSecretKeyRegex ="3F9IKGJRIT58TYRHhe93hg88skf95uf74hdy3=/d";

    @Test
    void matchRegex_empty() {
        assertFalse(SafeRegex.matchRegex(SafeRegex.safeTextNoSpacesRegex, ""));
        assertFalse(SafeRegex.matchRegex(SafeRegex.urlRegex, ""));
        assertFalse(SafeRegex.matchRegex(SafeRegex.uriRegex, ""));
        assertFalse(SafeRegex.matchRegex(SafeRegex.ipv4Regex, ""));
        assertFalse(SafeRegex.matchRegex(SafeRegex.urlOrIPV4Regex, ""));
        assertFalse(SafeRegex.matchRegex(SafeRegex.base64Regex, ""));
        assertFalse(SafeRegex.matchRegex(SafeRegex.awsAccessKeyRegex, ""));
        assertFalse(SafeRegex.matchRegex(SafeRegex.awsSecretKeyRegex, ""));
    }

    @Test
    void matchRegex_space() {
        assertFalse(SafeRegex.matchRegex(SafeRegex.safeTextNoSpacesRegex, " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.urlRegex, " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.uriRegex, " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.ipv4Regex, " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.urlOrIPV4Regex, " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.base64Regex, " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.awsAccessKeyRegex, " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.awsSecretKeyRegex, " "));
    }

    @Test
    void matchRegex_notAnchored() {

        assertTrue(SafeRegex.matchRegex(SafeRegex.safeTextNoSpacesRegex, exampleSafeTextNoSpacesRegex));
        assertTrue(SafeRegex.matchRegex(SafeRegex.urlRegex, exampleUrlRegex));
        assertTrue(SafeRegex.matchRegex(SafeRegex.uriRegex, exampleUriRegex));
        assertTrue(SafeRegex.matchRegex(SafeRegex.ipv4Regex, exampleIpv4Regex));
        assertTrue(SafeRegex.matchRegex(SafeRegex.urlOrIPV4Regex, exampleUrlRegex));
        assertTrue(SafeRegex.matchRegex(SafeRegex.urlOrIPV4Regex, exampleIpv4Regex));
        assertTrue(SafeRegex.matchRegex(SafeRegex.base64Regex, exampleBase64Regex));
        assertTrue(SafeRegex.matchRegex(SafeRegex.awsAccessKeyRegex, exampleAwsAccessKeyRegex));
        assertTrue(SafeRegex.matchRegex(SafeRegex.awsSecretKeyRegex, exampleAwsSecretKeyRegex));

        assertFalse(SafeRegex.matchRegex(SafeRegex.safeTextNoSpacesRegex, " " + exampleSafeTextNoSpacesRegex));
        assertFalse(SafeRegex.matchRegex(SafeRegex.urlRegex, " " + exampleUrlRegex));
        assertFalse(SafeRegex.matchRegex(SafeRegex.uriRegex, " " + exampleUriRegex));
        assertFalse(SafeRegex.matchRegex(SafeRegex.ipv4Regex, " " + exampleIpv4Regex));
        assertFalse(SafeRegex.matchRegex(SafeRegex.urlOrIPV4Regex, " " + exampleUrlRegex));
        assertFalse(SafeRegex.matchRegex(SafeRegex.urlOrIPV4Regex, " " + exampleUriRegex));
        assertFalse(SafeRegex.matchRegex(SafeRegex.base64Regex, " " + exampleBase64Regex));
        assertFalse(SafeRegex.matchRegex(SafeRegex.awsAccessKeyRegex, " " + exampleAwsAccessKeyRegex));
        assertFalse(SafeRegex.matchRegex(SafeRegex.awsSecretKeyRegex, " " + exampleAwsSecretKeyRegex));

        assertFalse(SafeRegex.matchRegex(SafeRegex.safeTextNoSpacesRegex, exampleSafeTextNoSpacesRegex + " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.urlRegex, exampleUrlRegex + " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.uriRegex, exampleUriRegex + " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.ipv4Regex, exampleIpv4Regex + " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.urlOrIPV4Regex, exampleUrlRegex + " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.urlOrIPV4Regex, exampleUriRegex + " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.base64Regex, exampleBase64Regex + " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.awsAccessKeyRegex, exampleAwsAccessKeyRegex + " "));
        assertFalse(SafeRegex.matchRegex(SafeRegex.awsSecretKeyRegex, exampleAwsSecretKeyRegex + " "));
    }
}

File: CryptoAC/src/commonTest/kotlin/eu/fbk/st/cryptoac/SafeRegexTest.kt
package eu.fbk.st.cryptoac

import kotlin.test.Test
import kotlin.test.assertFalse
import kotlin.test.assertTrue

internal class SafeRegexTest {

    private val exampleSafeTextNoSpacesRegex = "text"
    private val exampleUrlRegex = "https://google.com/"
    private val exampleUriRegex = "/#G1kiGui2qDYcdjriD0kdsspLycD75of"
    private val exampleIpv4Regex = "192.168.111.4"
    private val exampleBase64Regex = "At539hww"
    private val exampleAwsAccessKeyRegex = "A94KF94IR9FIEKDMCJ84"
    private val exampleAwsSecretKeyRegex ="3F9IKGJRIT58TYRHhe93hg88skf95uf74hdy3=/d"

    @Test
    fun `valid texts validate against regex`() {
        assertTrue(SafeRegex.TEXT.matches(exampleSafeTextNoSpacesRegex))
        assertTrue(SafeRegex.URL.matches(exampleUrlRegex))
        assertTrue(SafeRegex.URI.matches(exampleUriRegex))
        assertTrue(SafeRegex.IPV4.matches(exampleIpv4Regex))
        assertTrue(SafeRegex.URL_OR_IPV4.matches(exampleUrlRegex))
        assertTrue(SafeRegex.URL_OR_IPV4.matches(exampleIpv4Regex))
        assertTrue(SafeRegex.BASE64.matches(exampleBase64Regex))
        assertTrue(SafeRegex.AWS_ACCESS_KEY.matches(exampleAwsAccessKeyRegex))
        assertTrue(SafeRegex.AWS_SECRET_KEY.matches(exampleAwsSecretKeyRegex))
    }

    @Test
    fun `empty text do not validate against regex`() {
        assertFalse(SafeRegex.TEXT.matches(""))
        assertFalse(SafeRegex.URL.matches(""))
        assertFalse(SafeRegex.URI.matches(""))
        assertFalse(SafeRegex.IPV4.matches(""))
        assertFalse(SafeRegex.URL_OR_IPV4.matches(""))
        assertFalse(SafeRegex.BASE64.matches(""))
        assertFalse(SafeRegex.AWS_ACCESS_KEY.matches(""))
        assertFalse(SafeRegex.AWS_SECRET_KEY.matches(""))
    }

    @Test
    fun `blank text do not validate against regex`() {
        assertFalse(SafeRegex.TEXT.matches(" "))
        assertFalse(SafeRegex.URL.matches(" "))
        assertFalse(SafeRegex.URI.matches(" "))
        assertFalse(SafeRegex.IPV4.matches(" "))
        assertFalse(SafeRegex.URL_OR_IPV4.matches(" "))
        assertFalse(SafeRegex.BASE64.matches(" "))
        assertFalse(SafeRegex.AWS_ACCESS_KEY.matches(" "))
        assertFalse(SafeRegex.AWS_SECRET_KEY.matches(" "))
    }

    @Test
    fun `not anchored text at the beginning do not validate against regex`() {
        assertFalse(SafeRegex.TEXT.matches(" $exampleSafeTextNoSpacesRegex"))
        assertFalse(SafeRegex.URL.matches(" $exampleUrlRegex"))
        assertFalse(SafeRegex.URI.matches(" $exampleUriRegex"))
        assertFalse(SafeRegex.IPV4.matches(" $exampleIpv4Regex"))
        assertFalse(SafeRegex.URL_OR_IPV4.matches(" $exampleUrlRegex"))
        assertFalse(SafeRegex.URL_OR_IPV4.matches(" $exampleUriRegex"))
        assertFalse(SafeRegex.BASE64.matches(" $exampleBase64Regex"))
        assertFalse(SafeRegex.AWS_ACCESS_KEY.matches(" $exampleAwsAccessKeyRegex"))
        assertFalse(SafeRegex.AWS_SECRET_KEY.matches(" $exampleAwsSecretKeyRegex"))
    }

    @Test
    fun `not anchored text at the end do not validate against regex`() {
        assertFalse(SafeRegex.TEXT.matches("$exampleSafeTextNoSpacesRegex "))
        assertFalse(SafeRegex.URL.matches("$exampleUrlRegex "))
        assertFalse(SafeRegex.URI.matches("$exampleUriRegex "))
        assertFalse(SafeRegex.IPV4.matches("$exampleIpv4Regex "))
        assertFalse(SafeRegex.URL_OR_IPV4.matches("$exampleUrlRegex "))
        assertFalse(SafeRegex.URL_OR_IPV4.matches("$exampleUriRegex "))
        assertFalse(SafeRegex.BASE64.matches("$exampleBase64Regex "))
        assertFalse(SafeRegex.AWS_ACCESS_KEY.matches("$exampleAwsAccessKeyRegex "))
        assertFalse(SafeRegex.AWS_SECRET_KEY.matches("$exampleAwsSecretKeyRegex "))
    }
}